src/lua/version.lua

```diff
@@ -0,0 +1,259 @@
+--
+-- Version parser adopted to the Tarantool's versioning policy:
+-- https://www.tarantool.io/en/doc/latest/release/policy/#versioning-policy
+--
+-- Tarantool's versions use SemVer-like format:
+--     x.y.z-typeN-commit-ghash
+--
+-- * x.y.z - major, minor, patch release numbers;
+-- * typeN - pre-release name and its optional number: alpha1, rc10. Optional;
+-- * commit - number of commits since the latest release. Optional;
+-- * ghash - latest commit hash in format g<hash>. Optional.
+--
+-- Differences with the semver docs:
+--
+-- * Pre-release identifiers are not separated. Not 'alpha.1' but 'alpha1'.
+-- * Pre-release identifiers cannot be nested. Not x.y.z-alpha.beta.
+-- * Pre-release identifier always includes non-numeric type: e.g. beta, rc.
+-- * Build metadata is denoted by appending hyphen, not plus sign. Not
+--   x.y.z-typeN+commit-ghash, but x.y.z-typeN-commit-ghash.
+--
+-- Note, that ghash itself and everything after it (e.g. -dev suffix) is
+-- ignored and doesn't affect comparing of the versions.
+--
+
+local ffi   = require('ffi')
+local utils = require('internal.utils')
+
+local check_param = utils.check_param
+
+-- Weight of the empty prerelease name.
+local MAX_PRERELEASE_NAME_WEIGHT = 255
+
+--
+-- Cdata is used in order to allow comparison of the version object with
+-- strings. Since relational metamethods do not support mixed types in Lua,
+-- using tables is not possible.
+--
+ffi.cdef([[
+    struct version {
+        /** Major number. */
+        uint32_t major;
+        /** Minor number. */
+        uint32_t minor;
+        /** Patch number. */
+        uint32_t patch;
+        /** Pre-release name id. */
+        uint8_t _prerelease_id;
+        /** Pre-release number: e.g. in alpha1 it's 1. */
+        uint16_t _prerelease_num;
+        /** Number of commits since release. */
+        uint32_t commit;
+    };
+]])
+
+--
+-- Pre-release names sorted in descending order.
+-- ID 0 is reserved for missing pre-release name.
+--
+local prerelease_names = {
+    'rc',
+    'beta',
+    'alpha',
+    'entrypoint',
+}
+
+local prerelease_name_to_id = {}
+for id, name in ipairs(prerelease_names) do
+    prerelease_name_to_id[name] = id
+end
+
+local function version_new(major, minor, patch, prerelease, commit)
+    commit = commit or 0
+    check_param(major, 'major', 'number', 2)
+    check_param(minor, 'minor', 'number', 2)
+    check_param(patch, 'patch', 'number', 2)
+    check_param(commit, 'commit', 'number', 2)
+
+    local prerelease_id
+    local prerelease_num = 0
+    if prerelease then
+        local prerelease_name
+        check_param(prerelease, 'prerelease', 'string', 2)
+        prerelease_name, prerelease_num = prerelease:match('^(%a+)(%d+)')
```

Comment:
Here we also accept something unexpected:

```lua
tarantool> version.new(3, 3, 0, 'rc5.pre1')
---
- 3.3.0-rc5
...
```

Reply:
See https://github.com/tarantool/tarantool/pull/10717#discussion_r1855700695